haystack/document_stores/in_memory/document_store.py

```diff
@@ -50,15 +51,200 @@ def __init__(
                                               To choose the most appropriate function, look for information about your embedding model.
         """
         self.storage: Dict[str, Document] = {}
-        self._bm25_tokenization_regex = bm25_tokenization_regex
+        self.bm25_tokenization_regex = bm25_tokenization_regex
         self.tokenizer = re.compile(bm25_tokenization_regex).findall
-        algorithm_class = getattr(rank_bm25, bm25_algorithm)
-        if algorithm_class is None:
-            raise ValueError(f"BM25 algorithm '{bm25_algorithm}' not found.")
-        self.bm25_algorithm = algorithm_class
+
+        self.bm25_algorithm = bm25_algorithm
+        self.bm25_algorithm_inst = self._dispatch_bm25()
         self.bm25_parameters = bm25_parameters or {}
         self.embedding_similarity_function = embedding_similarity_function
 
+        # Global BM25 statistics
+        self._avg_doc_len: float = 0.0
+        self._freq_doc: Counter = Counter()
+
+        # Per-document statistics
+        self._bm25_attr: Dict[str, Tuple[Dict[str, int], int]] = {}
```

Comment:
I think we should have the per doc stats represented in a more structured way, e.g.:

```
@dataclass
class DocumentStats:
    """
    Document statistics for BM25 scoring.
    """
    freq: Counter[str]
    doc_len: int
```
also adding any methods one might need. 

Then, we could define this structure like this:

```
# Per-document statistics
self._bm25_attr: Dict[str, DocumentStats] = {}

```

Reply:
I agree, will explore this option :)

Reply:
Updated in [this commit](https://github.com/deepset-ai/haystack/pull/7549/commits/687bef3b51145400058c746734217cf5b59b132d) as `BM25DocumentStats` data class.